chain/x/zoracle/internal/keeper/report.go

```diff
@@ -6,6 +6,68 @@ import (
 	"github.com/bandprotocol/d3n/chain/x/zoracle/internal/types"
 )
 
+func (k Keeper) AddReport(
+	ctx sdk.Context, requestID int64, dataSet []types.RawDataReport, validator sdk.ValAddress,
+) sdk.Error {
+	request, err := k.GetRequest(ctx, requestID)
+	if err != nil {
+		return err
+	}
+
+	if request.IsResolved || request.ExpirationHeight < ctx.BlockHeight() {
+		// TODO: fix error later
+		return types.ErrRequestNotFound(types.DefaultCodespace)
+	}
+
+	found := false
+	for _, validValidator := range request.RequestedValidators {
+		if validator.Equals(validValidator) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return types.ErrInvalidValidator(types.DefaultCodespace)
+	}
+
+	isFirstReport := true
```

Comment:
This is complicated. Let's simplify by requiring that all raw data reports must come in one message.